### PR TITLE
Add Njalla provider

### PR DESCRIPTION
**Pull request by chapatt**
This implements Njalla's new API: https://njal.la/api/

### CODE REVIEW

File: lexicon/providers/njalla.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            'domain': self.domain,
+            'id': identifier,
+        }
+        self._api_call('remove-record', params)
+
+        LOGGER.debug('delete_record: %s', True)
+        return True
+
+    # Helpers
+    def _api_call(self, method, params):
```
Comment (adferrand) [inline lexicon/providers/njalla.py:113]: Do you have a specific reason to not implement the method `_request` from the base provider? It seems to me that it would fit your case, and you could call `_post` in place of `_api_call`.

File: lexicon/providers/njalla.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            raise Exception('%d: %s' % (error['code'], error['message']))
+
+        return response['result']
+
+    def _get_record_identifier(self, rtype=None, name=None, content=None):
+        records = self._list_records(rtype=rtype, name=name, content=content)
+        if len(records) == 1:
+            return records[0]['id']
+
+        raise Exception('Unambiguous record could not be found.')
```
Comment (adferrand) [inline lexicon/providers/njalla.py:142]: Just to inform you here. You have basically two choices if more than one record is found:
* error out, like you did
* get the first matching record, and warn the user that more than a record was found, and so the first one will be taken

Most of the providers use the second option, but it is your call.

File: lexicon/providers/njalla.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    def _request(self, action='GET', url='/', data=None, query_params=None):
```
Comment (adferrand) [inline lexicon/providers/njalla.py:144]: See my comment above, about `_api_call`.

File: lexicon/tests/providers/test_njalla.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class NjallaProviderTests(TestCase, IntegrationTests):
+    """TestCase for Njalla"""
+    provider_name = 'njalla'
+    domain = 'example.com'
+
+    def _filter_headers(self):
+        return ['Authorization']
+
+    @pytest.fixture(autouse=True)
+    def _skip_suite(self, request):  # pylint: disable=no-self-use
```
Comment (adferrand) [inline lexicon/tests/providers/test_njalla.py:20]: Are you willing to implement the advanced test suite? It is mainly about the capacity to handle multiple records for the same name (like TXTs). It is mostly useful for ACME clients, like Certbot, to issue wildcard certificates, which require two token for the same name.

It is not mandatory, but I wanted to give you an insight about it.

File: lexicon/providers/njalla.py (hunk 1 above)
Comment (chapatt) [reply lexicon/providers/njalla.py:113]: I expected a comment on this. It just seemed weird to implement the `"request` method when the JSON-RPC API only takes POSTs. If I did, I'd still be tempted to wrap `_post` in another method that takes `method` and `params` args.

I just realized that the API docs are even require login, so some of that might not have been apparent. I'm willing to use `_request` if you feel that's important.

File: lexicon/providers/njalla.py (hunk 2 above)
Comment (chapatt) [reply lexicon/providers/njalla.py:142]: I'm using this method in `_delete_record` so deleting the first of multiple matched records seems dangerous.

File: lexicon/tests/providers/test_njalla.py (hunk 4 above)
Comment (chapatt) [reply lexicon/tests/providers/test_njalla.py:20]: I ran into an issue with the very first one: `test_provider_when_calling_create_record_with_duplicate_records_should_be_noop`
- This is not the native behavior of Njalla (it allows duplicate records)
- I attempted doing the check myself with a `_list_records` in the beginning of `_create_record`, but it appeared to not return results for the duplicate record created only a moment before. It seems operations aren't synchronous, and it takes a while for the system to become consistent.

As completely duplicate records are permitted, record sets are also, and are returned correctly by `_list_records` with filter parameters.

I have however not implemented `def test_provider_when_calling_delete_record_with_record_set_name_remove_all`. Is deleting record sets in one go required for by certbot?

File: lexicon/providers/njalla.py (hunk 1 above)
Comment (adferrand) [reply lexicon/providers/njalla.py:113]: I understand that for this provider specifically passing through `_request` may be a burden, but it is for a more general purpose. You see, Lexicon is composed of variety of providers developped by a lot of different people. At some point of Lexicon lifecycle, it is necessary to make some global modifications.

Having some "standard" APIs help a lot to avoid to miss something during these kind of refactoring. In particular, having the actual network requests passing through `_request` is helping to avoid missing specific logic if a global refactoring impact the network operations.

And I am completely ok to wrap the underlying `_request` calls to more suitable high level functions for the Njalla provider.

File: lexicon/providers/njalla.py (hunk 2 above)
Comment (adferrand) [reply lexicon/providers/njalla.py:142]: I agree. My option makes more sense in update operation, but for deleting it is bad.

File: lexicon/tests/providers/test_njalla.py (hunk 4 above)
Comment (adferrand) [reply lexicon/tests/providers/test_njalla.py:20]: Alright, may be tricky to match the expected behavior for the advanced test suite. No, deleting all the record set is not mandatory: Certbot will remove a specific TXT+entry+value for each related challenge in a given certificate.

**Comment by adferrand** [comment]
Really sorry for the late feedback. Appart `_request`, it is almost done! Don't mind the failure in circle ci. By the way, if you merge master, you will see it disappear.

**Comment by adferrand** [comment]
Don't bother the merge from master, I did it for you.

**Comment by adferrand** [comment]
Hello @chapatt!

Are you willing to finish the PR? Thanks in advance!

File: lexicon/providers/njalla.py (hunk 1 above)
Comment (chapatt) [reply lexicon/providers/njalla.py:113]: I implemented `_request` :)

**Comment by chapatt** [comment]
Sorry, I lost this PR for a bit. Please let me know if there's anything else that needs fixing up.

**Comment by chapatt** [comment]
@adferrand I see I'm failing all of the automated tests now. I can't quite make sense of the errors.

After merging in master, the developer guide local test procedure doesn't seem to work anymore:
```
virtualenv -p python2.7 venv
source venv/bin/activate
py.test lexicon/tests/providers/test_njalla.py
```

**Comment by adferrand** [comment]
Hello @chapatt! Really sorry for the delay, I was on vacations. Indeed I changed a little the design of integration tests.

Before you were extending your test class from `IntegrationTest`, with optional explicit markers to disable some parts of the test suite. Now the actual test suite run is defined by the class itself: `IntegrationTestsV2` or `IntegrationTestsV1`. You should use `IntegrationTestsV2` if possible, because the test suite is more complete.

So basically, replace `IntegrationTests` by `IntegrationTestsV2` in your test class, and you are good.

**Comment by chapatt** [comment]
> Hello @chapatt! Really sorry for the delay, I was on vacations. Indeed I changed a little the design of integration tests.
> 
> Before you were extending your test class from `IntegrationTest`, with optional explicit markers to disable some parts of the test suite. Now the actual test suite run is defined by the class itself: `IntegrationTestsV2` or `IntegrationTestsV1`. You should use `IntegrationTestsV2` if possible, because the test suite is more complete.
> 
> So basically, replace `IntegrationTests` by `IntegrationTestsV2` in your test class, and you are good.

No worries! I updated to `IntegrationTestsV2`. I still skipped a couple tests because Njalla is not natively capable of them. If they're necessary, it will have to be done client side by querying first, then modifying behavior based on that.

It also looks like I'm failing a couple checks. I can't seem to find useful output of those builds to see why.
